### PR TITLE
DSDEEPB-1500 transition attributes to main summary screen

### PR DESCRIPTION
**Pull request by eddiebroad**
…rking

load in the attributes to the state here

wip debug

some WIP

some WIP

cleanup and prep for saving and adjusting focus

got buttons for unrelated things to disapper
got focus working
next to work on updating and deleting

some WIP

need to get update working

got both deleting and updating to work, now focus on 1 tidy , 2 input validation 3 blockers

added some input validation ; now address docker and tidy

add blockers

made error handling a bit more graceful

added uniq check to keys

remove unused functions

### CODE REVIEW

**Comment by MattPutnam** [comment]
Cool, all combinations of editing/deleting/canceling that I could come up with worked perfectly.

There's a usability issue where the attributes are way down the page, but the edit and save/cancel buttons are at the top, forcing the user to scroll around.  There was some earlier discussion about reorganizing the various things to display, and now would be a good time to look at that closer.  We could make a second column (like the submission details page), we could do the floating sidebar (like the method config editor), or maybe both.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs [hunk 1]
```diff
@@ -13,141 +13,65 @@
...
-                 (if success?
-                   (let [response (get-parsed-response)
-                         ws (response "workspace")
-                         attrs-list (mapv (fn [[k v]] [k v]) (ws "attributes"))]
-                     (swap! state assoc :server-response
-                       {:attrs-list attrs-list} :loaded-attrs? true :saving? true)
-                     (on-success))
-                   (swap! state assoc :server-response {:error-message status-text})))}))
+(defn view-attributes [state refs]
+  [:div {:style {:margin "45px 25px"}}
```
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs]: This is causing the attributes section to be indented over weirdly.  There are a lot of nested divs here--are they all necessary?

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs [hunk 2]
```diff
@@ -13,141 +13,65 @@
...
+                                   (swap! state update-in [:attrs-list] conj ["" ""])
+                                   (js/setTimeout
+                                        (fn []
+                                          (common/focus-and-select
+                                            (-> (@refs (str "key_" (dec (count (:attrs-list @state))))) .getDOMNode)))0))}]]])
+     [:div {:style {:display
+                    (when (or (or (:editing? @state)
+                                (not-empty (:attrs-list @state))) (:saving? @state)) "none")}}
+      (style/create-paragraph [:em {} "There are no attributes to display"])]]]
+   (common/clear-both)])
```
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs]: `clear-both` is unnecessary here.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs (hunk 1 above)
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs]: On closer inspection, no.  These 3 divs can collapse into a single one.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs [hunk 3]
```diff
@@ -53,23 +60,104 @@
...
+      (when-not editing?
+        [comps/SidebarButton {:style :light :margin :top :color :button-blue
                             :text "Clone..." :icon :plus
-                            :onClick #(swap! state assoc :cloning? true)}]
+                            :onClick #(swap! state assoc :cloning? true)}])
       (when owner?
-        [comps/SidebarButton {:style :light :margin :top :color :button-blue
-                              :text (if locked? "Unlock" "Lock") :icon :locked
-                              :onClick #(react/call :lock-or-unlock this locked?)}])
+        (if editing? nil
```
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: `(when (and owner? (not editing?))`

Similar below

**Comment by eddiebroad** [comment]
I guess you saw that in the beginning of the code I sort of "make a restore point" of the attrs-list.  In the event of cancel, that's restored.  The thing that was hard to test was recovering from errors,....this is not totally dissimilar from some stuff you mentioned at stand-up this morning.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs [hunk 4]
```diff
@@ -53,23 +60,104 @@
...
+           {:style :light :color :button-blue :margin :top
+            :text "Edit attributes" :icon :pencil
+            :onClick #(swap! state assoc
+                       :reserved-keys (vec (range 0 (count (:attrs-list @state))))
+                       :orig-attrs (:attrs-list @state) :editing? true)}]
+          [:div {}
+           [comps/SidebarButton
+            {:style :light :color :button-blue :margin :top
+             :text "Save Attributes" :icon :document
+             :onClick #(do
```
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: This `do` is unnecessary.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs [hunk 5]
```diff
@@ -53,23 +60,104 @@
...
+                                                          (swap! state dissoc :deleting-attrs?)
+                                                          (if-not success?
+                                                            (do
+                                                              (js/alert (str "Exception:\n" (.-statusText xhr)))
+                                                              (swap! state assoc :attrs-list (:orig-attrs @state))
+                                                              (swap! state dissoc :orig-attrs)
+                                                              (react/call :load-workspace this))
+                                                            (when-not (zero? (count up-mapv))
+                                                              (update-orch-fn  up-mapv))))}))
+                              uniq-keys (== (count curr-keys) (count (distinct curr-keys)))]
```
Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: `(apply distinct? curr-keys)`

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs (hunk 1 above)
Comment (eddiebroad) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/attribute_editor.cljs]: Re: "On closer inspection...." It looks like you answered your own question.  I will collapse them into a single one.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs [hunk 6]
```diff
@@ -139,10 +219,21 @@
...
                        (if success?
-                         {:workspace (get-parsed-response)}
-                         {:workspace-error status-text})
+                         (let [response (:server-response @state)
+                               r1 (utils/rlog (str "the response is " response))
+                               attributes (get-in response [:workspace "workspace" "attributes" ])
+                               r2 (utils/rlog (str "attrs is " attributes))
+                               attrs-list (mapv (fn [[k v]] [k v]) attributes)]
+                           (utils/rlog (str "the response is " response))
+                           (utils/rlog (str "the attrs-list is "attrs-list))
```
Comment (eddiebroad) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: Re "Take out logging calls"  I had them in there for so long I didn't notice them anymore.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs (hunk 5 above)
Comment (eddiebroad) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: do you mean "uniq-keys (apply distinct? curr-keys)]" ?

Comment (MattPutnam) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/tab.cljs]: Right

**Comment by MattPutnam** [comment]
I'm giving this a :+1:.  There's still some weirdness with the "description" attribute, but I'm going to work on editing the description soon anyway.

**Pull request merged**